cdap-watchdog/src/main/java/co/cask/cdap/logging/write/AvroFileWriter.java

```diff
@@ -226,44 +237,74 @@ public AvroFile(Location location) {
     }
 
     /**
-     * Opens the underlying file for writing. If open throws an exception, then @{link #close()} needs to be called to
-     * free resources.
+     * Opens the underlying file for writing.
+     * If open throws an exception then underlying file may still need to be deleted.
+     *
      * @throws IOException
      */
     void open() throws IOException {
-      this.outputStream = new FSDataOutputStream(location.getOutputStream(), null);
-      this.dataFileWriter = new DataFileWriter<>(new GenericDatumWriter<GenericRecord>(schema));
-      this.dataFileWriter.create(schema, this.outputStream);
-      this.dataFileWriter.setSyncInterval(syncIntervalBytes);
-      this.lastModifiedTs = System.currentTimeMillis();
+      try {
+        this.outputStream = new FSDataOutputStream(location.getOutputStream(), null);
+        this.dataFileWriter = new DataFileWriter<>(new GenericDatumWriter<GenericRecord>(schema));
+        this.dataFileWriter.create(schema, this.outputStream);
+        this.dataFileWriter.setSyncInterval(syncIntervalBytes);
+        this.lastModifiedTs = System.currentTimeMillis();
+      } catch (Exception e) {
+        close();
+        throw e;
+      }
       this.isOpen = true;
     }
 
+    public boolean isOpen() {
+      return isOpen;
+    }
+
     public Location getLocation() {
       return location;
     }
 
     public void append(LogWriteEvent event) throws IOException {
-      dataFileWriter.append(event.getGenericRecord());
-      lastModifiedTs = System.currentTimeMillis();
+      try {
+        dataFileWriter.append(event.getGenericRecord());
+        lastModifiedTs = System.currentTimeMillis();
+      } catch (Exception e) {
+        close();
+        throw e;
```

Comment:
I had the same thought when I did that, but surprisingly it worked. It is probably because the only checked exception that is thrown in the block is IOException.